### DISCUSSION TITLE
Mixing callbag-jsx with other JSX-based libraries

### DISCUSSION THREAD

**Discussion by rosenfeld** [loreanvictor/callbag-jsx#12] [Q&A]
I'm sorry if this is not the proper place to ask this, but I couldn't find a mailing list related to this very interesting project.

I'm currently mixing React in an old existing jQuery based app, which also has some part written in Knockout.js. Callback-JSX seems way more interesting to me to integrate into this project than React, since I can mix the callback-jsx components with manually interacting with the DOM (through jQuery components, for example).

However, I'd love to use the JSX approach with callbag-jsx too, but Babel is configured to use JSX with React. Are there any resources on what approach could allow me to configure JSX depending on the source file? Maybe I could use some different JS extension to differentiate between React and Callbag-JSX sources, or maybe some magical comments in the sources. I realize this is probably a question for Babel JSX related plugins, but maybe this is something other people interested in callbag-JSX could be concerned about. If there is an easy way to mix React and Callbag-JSX while using the JSX approach for both, it would be awesome to have a section explaining how to configure that in Babel and webpack, for example.

Anyway, I'm very excited about callbag-JSX, congrats for the new approach!

It looks to me like a great mix of React and Knockout.js features. Loved it. Thanks for callbag-jsx, I think the ecosystem was lacking a library with that approach, well done!

**Comment by loreanvictor** [comment]
Thanks a lot!

[Here](https://loreanvictor.github.io/callbag-jsx/install) I tried to explain how to use Callbag-JSX with Babel or TypeScript. To summarize:

- If you use one of the [starter templates](https://loreanvictor.github.io/callbag-jsx/install#starter-templates), they are pre-configured to properly transpile JSX.
- If you want to [manually install Callbag-JSX](https://loreanvictor.github.io/callbag-jsx/install), you have two options:
  1. Configuring your transpiler project-wide. For Babel, this would look like this:
  ```js
  // .babelrc
  {
    "plugins": [
       ["@babel/plugin-transform-react-jsx", {
         "pragma": "renderer.create",
         "pragmaFrag": "renderer.fragment"
       }]
    ]
  }
  ```
  2. Configuring JSX transpilation on a per-file basis, i.e. add these pragmas at top of the files:
  ```js
  /** @jsx renderer.create */
  /** @jsxFrag renderer.fragment */
  ```
  
You don't need any particular configuration for Webpack as Webpack will fallback to your installed transpiler for transpilation, so configuring Typescript or Babel should suffice.
  
For other bundler / transpilers, it really depends on the case. For example, if you want to use [Snowpack](https://www.snowpack.dev) (which I would highly recommend btw, as I'll also migrate the starter templates to use Snowpack instead of Webpack soon), then the following Snowpack configuration would do the trick:

```js
// snowpack.config.js
module.exports = {
  /* ... */
  buildOptions: {
    /* ... */
    jsxFactory: 'renderer.create',
    jsxFragment: 'renderer.fragment',
  },
}
```

Hope that helps!

P.S. if this answers your question, please ping me so that I'd move this to **Disucssions/Q&A**. If it doesn't, also don't hesitate to ping me, as it might mean something is missing from the docs that needs to be added!

**Comment by loreanvictor** [comment]
> but I couldn't find a mailing list related to this very interesting project.

Also good point! I do realize that more casual formats can be more welcoming for people compared to **Issues** / **Discussions** sections on GitHub. I will make a Gitter room and link it on the docs.

**Comment by rosenfeld** [comment]
This JSX transpilation pragmas are exactly what I was looking for, thanks! I'll give it a time as soon as I have some free time. I need even more time to check out Snowpack, since we do have some custom webpack plugins in place and it may not be so simple to replace webpack without a major planning.

Thanks a lot! I'll try the Gitter room next time I need some help :)

**Comment by rosenfeld** [comment]
While there isn't yet a communication project for callbag-jsx, please allow me to ask more questions here.

Are there any plans to support classes as components when rendering the JSX? All examples seem to use a function that gets the props and renderer and returns what should be rendered. Have you considered supporting classes as well? For example, if the class has a method "render" it could be called with the props and renderer arguments, maybe. Some people might prefer classes over function for organizing their sources...

This is not a big deal, since it should be easy to create a wrapper for such classes (so that we could do something like this after defining the class `const ComponentRenderer = wrapper(ComponentClass);`, but it might be interesting to support it out of the box...

**Comment by loreanvictor** [comment]
@rosenfeld I created a separate issue for your question, since this way other people facing the same questions will be able to find them easier.
